cpp/src/binaryop/compiled/operation.cuh

```diff
@@ -91,11 +91,46 @@ struct TrueDiv {
 };
 
 struct FloorDiv {
-  template <typename T1, typename T2>
-  __device__ inline auto operator()(T1 const& lhs, T2 const& rhs)
-    -> decltype(floor(static_cast<double>(lhs) / static_cast<double>(rhs)))
+  template <typename TypeLhs,
+            typename TypeRhs,
+            std::enable_if_t<(std::is_integral_v<std::common_type_t<TypeLhs, TypeRhs>> and
+                              std::is_signed_v<std::common_type_t<TypeLhs, TypeRhs>>)>* = nullptr>
+  __device__ inline auto operator()(TypeLhs x, TypeRhs y) -> decltype(x / y)
+  {
+    if ((x ^ y) >= 0) {
+      return x / y;
+    } else {
+      const auto quotient  = x / y;
+      const auto remainder = x % y;
```

Comment:
It might be better to implement this branch-free:

```c++
auto const quotient = x / y;
auto const nonzero_remainder = !!(x % y);
auto const mixed_sign = (x ^ y) < 0;
return quotient - mixed_sign * nonzero_remainder;
```

Reply:
It was.